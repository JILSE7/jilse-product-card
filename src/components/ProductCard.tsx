import styles from '../styles/styles.module.css';


import React, { createContext } from 'react';
import useCount from '../hooks/useCount';
import { IProductContext, IPropsProductCardComponent } from '../interfaces/ProductInterfaces';



export const ProductContext = createContext({} as IProductContext);
const {Provider} = ProductContext;

export const ProductCard = ({product, children, className, style, onChange, value, initialValues}:IPropsProductCardComponent) => {
    
  

  const {counter, increasyBy, maxCount, isMaxCountReached, reset} = useCount({onChange, product,value, initialValues});

  return (
    <Provider value={{counter, increasyBy, product, maxCount}}>
      
        <div className={`${styles.productCard} ${className}`} style={style}>

            {
           
                  children({count:counter,increasyBy,isMaxCountReached,maxCount: maxCount!, reset, product})
               
            }

        </div>

    </Provider>
  )
}

//*OTRA FORMA DE EXPORTAR LOS COMPONENTES PARA QUE SE MIREN DE LA SIGUIENTE FORMA
//?PASAMOS DE ESTO, QUE NO ESTA MAL, SOLO ES UNA FORMA 
{/* <ProductCard product={product}>
  <ProductImage/>
  <ProductTitle title="Coffe Cup"/>
  <ProducButtons counter={counter} increasyBy={increasyBy}/>
</ProductCard> */}

//?PASAMOS A PONERLE MAS PROPIEDADES AL COMPONENTE
 //* <ProductCard product={product}>
      //?<ProductCard.Image/>
      //?<ProductCard.Title title="Coffe Cup"/>
      //?<ProductCard.Buttons counter={counter} increasyBy={increasyBy}/>
//* </ProductCard> 

//*ASIGNAMOS LAS PROPIEDADES


