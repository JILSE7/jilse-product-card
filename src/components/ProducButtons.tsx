import React,{ useCallback, useContext } from "react";
import { ProductContext } from "./ProductCard";
import styles from '../styles/styles.module.css';
import { IProductButtonsProps } from "../interfaces/ProductInterfaces";


export const ProducButtons = ({className, style}:IProductButtonsProps) => {
    const {counter, increasyBy, maxCount} = useContext(ProductContext);

    const maxCountReaced = useCallback(
      () => !!maxCount && counter === maxCount ,
      [counter,maxCount],
    )
    
    console.log(!!maxCount);

  return(
        <div className={`${styles.buttonsContainer} ${className}`} style={style}>
          <button className={styles.buttonMinus} onClick={() => increasyBy(-1)}>-</button>

          <div className={styles.countLabel}>{counter}</div>

          <button className={`${styles.buttonAdd} ${maxCountReaced() && styles.disabled}`}  onClick={() => increasyBy(1)}>+</button>
        </div>
  )
}


//other solution

/* const maxCountReaced = useCallback(
  () => {
    if(counter === maxCount) return true;
    else return false;
  },
  [counter,maxCount],
) */
