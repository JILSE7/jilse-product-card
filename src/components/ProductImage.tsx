import React,{ useContext } from 'react';
import { ProductContext } from './ProductCard';
import { IProductImageProps } from '../interfaces/ProductInterfaces';
import styles from '../styles/styles.module.css';
import noImage from '../assets/no-image.jpg';


export const ProductImage = ({img = '', className, style}:IProductImageProps) => {

    const {product} = useContext(ProductContext);
  
    return (
      <img className={`${styles.productImg} ${className}`} src={(img) ? img : product.img ?? noImage} alt="Coffe mug" style={style}/>
    )
  }