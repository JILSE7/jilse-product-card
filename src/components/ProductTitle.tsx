import React,{ useContext } from 'react';
import { ProductContext } from './ProductCard';
import styles from '../styles/styles.module.css';
import { IProductTitleProps } from '../interfaces/ProductInterfaces';




export const ProductTitle = ({title = '', className, style}: IProductTitleProps) => {
    const {product} = useContext(ProductContext);
    return (
      <span className={`${styles.productDescription} ${className}`} style={style}>{title ? title : product.nameProduct}</span>
    )
  }
  