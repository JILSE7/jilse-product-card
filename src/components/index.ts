import {ProductCard as ProductCardHOC} from './ProductCard';
import { ProducButtons } from './ProducButtons';
import { ProductImage } from './ProductImage';
import { ProductTitle } from './ProductTitle';
import { ProductCardHOCProps } from '../interfaces/ProductInterfaces';


export{
    ProducButtons,
    ProductImage,
    ProductTitle
}

const ProductCard:ProductCardHOCProps = Object.assign(ProductCardHOC, {
    Title: ProductTitle,
    Image: ProductImage,
    Buttons: ProducButtons,
    
})

export default ProductCard;