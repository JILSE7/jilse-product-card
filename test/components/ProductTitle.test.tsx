import React from 'react';
//import * as ReactDOM from 'react-dom';
import rendered from 'react-test-renderer';
import { ProductCard } from '../../src/components/ProductCard';
import { ProductTitle } from '../../src/components/ProductTitle';
import { product } from '../data/products';




describe('Pruebas en el product title', () => {

    test('should show component with personalize title', () => { 
        
        const wrapper = rendered.create(
            <ProductTitle title='Custom Product' className='said'/>
        )

        expect(wrapper.toJSON()).toMatchSnapshot();

    })

    test('debe de mostrar el componente con el nombre del producto', () => {
          
        const wrapper = rendered.create(
            <ProductCard product={product}>
                {
                    () => (
                        <>
                              <ProductTitle title='Custom Product' className='said'/>
                        </>
                    )
                }
            </ProductCard>
        )

        expect(wrapper.toJSON()).toMatchSnapshot();
    })
})
