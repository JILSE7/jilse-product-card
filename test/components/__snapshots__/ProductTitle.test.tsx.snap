// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el product title debe de mostrar el componente con el nombre del producto 1`] = `
<div
  className="productCard undefined"
>
  <span
    className="productDescription said"
  >
    Custom Product
  </span>
</div>
`;

exports[`Pruebas en el product title should show component with personalize title 1`] = `
<span
  className="productDescription said"
>
  Custom Product
</span>
`;
